File: solutions/109.Convert_Sorted_List_to_Binary_Search_Tree/AC_inorder_n.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_inorder_n.py
# Create Date: 2015-03-09 11:35:57
# Usage:       AC_inorder_n.py 
# Descripton:  


# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:

    def dfs(self, left, right):
        if left > right:
            return None

        mid = (left + right) // 2
        left_sub_tree = self.dfs(left, mid - 1)  # calc the left tree first

        root = TreeNode(self.head.val)
        self.head = self.head.next

        root.left = left_sub_tree
        root.right = self.dfs(mid + 1, right)
        return root

    # @param head, a list node
    # @return a tree node
    def sortedListToBST(self, head):
        cur = head

        # calc the size
        size = 0
        while cur:
            cur = cur.next
            size += 1

        self.head = head
        return self.dfs(0, size - 1)


# debug
s = Solution()
a = ListNode(1)
print s.sortedListToBST(a)
a.next = ListNode(2)
print s.sortedListToBST(a)
a.next.next = ListNode(3)
print s.sortedListToBST(a)

File: solutions/114.Flatten_Binary_Tree_to_Linked_List/AC_in_order_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_n.cpp
*  Create Date: 2015-03-10 14:27:07
*  Descripton:  Dfs, recursive and combind.
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    void flatten(TreeNode *root) {
        if (!root)
            return;

        TreeNode *left = root->left, *right = root->right;

        // recursive
        flatten(left);
        flatten(right);
        
        // combind
        root->left = NULL;
        root->right = left;
        while (root->right)
            root = root->right;
        root->right = right;
    }
};

int main() {

    return 0;
}

